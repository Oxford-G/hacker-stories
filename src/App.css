.container {
  height: 100vw;
  padding: 20px;
  background: #83a4d4; /* fallback for old browsers */
  background: linear-gradient(to left, #b6fbff, #83a4d4);
  color: #171212;
}

.headline-primary {
  font-size: 48px;
  font-weight: 300;
  letter-spacing: 2px;
}

.item {
  display: flex;
  align-items: center;
  padding-bottom: 5px;
}

.item > span {
  padding: 0 5px;
  white-space: nowrap;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.item > span > a {
  color: inherit;
}
