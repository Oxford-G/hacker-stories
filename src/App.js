import * as React from 'react';
import axios from 'axios';

const welcome = {
  greetings: 'Hey',
  title: 'React'
}

/*const initialStories = [
  {  title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  },

  {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
  }
];*/

/*const getAsyncStories = () => 
  new Promise((resolve) => 
    setTimeout(()=> resolve({data:{stories: initialStories}}), 2000)
  );*/

const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';

const useSemiPersistentState = (key, initialState) => {
  const [value, setValue] = React.useState(localStorage.getItem(key) || initialState);

  React.useEffect(() => {localStorage.setItem(key, value)}, [value, key])

  return [value, setValue]
}

const storiesReducer = (state, action) => {
  switch(action.type) {
    case 'STORIES_FETCH_INIT':
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case 'STORIES_FETCH_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload
      };
    case 'STORIES_FETCH_FAILURE':
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case 'REMOVE_STORY':
      return {
        ...state,
        data: state.data.filter(
        (story) => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();
  }
};

const App = () => {

  const[searchTerm, setSearchTerm] = useSemiPersistentState('search', 'React');

  const [url, setUrl] = React.useState(
    `${API_ENDPOINT}${searchTerm}`
  );

  // const [stories, setStories] = React.useState([]);

  const [stories, dispatchStories] = React.useReducer(storiesReducer,
    { data: [], isLoading: false, isError: false });

  /*const [isLoading, setIsLoading] = React.useState(false);*/

  /*const [isError, setIsError] = React.useState(false)*/

  const handleFetchStories = React.useCallback(()=>{
    if (searchTerm === '') return;

    /*setIsLoading(true);*/
    dispatchStories({ type: 'STORIES_FETCH_INIT' });
    
    // getAsyncStories()

    // fetch(`${API_ENDPOINT}${searchTerm}`).then(response => response.json())

    // fetch(`${url}`).then(response => response.json())

    // axios.get(url)
    // .then((result) => {

      // setStories(result.data.stories);

      // dispatchStories({
        // type: 'STORIES_FETCH_SUCCESS',
        // payload: result.data.stories
        // i am using data since axios wraps the result into a data object in JavaScript
        // payload: result.data.hits
      // });

      // setIsLoading(false);
    // })
    // .catch(()=> dispatchStories({ type: 'STORIES_FETCH_FAILURE' }))
    try{
      const result = await axios.get(url);
      dispatchStories({
        type: 'STORIES_FETCH_SUCCESS',
        payload: result.data.hits,
        })
    } catch{dispatchStories({ type: 'STORIES_FETCH_FAILURE' })};
  }, [url]);
  
  React.useEffect(() =>{
    handleFetchStories();
  }, [handleFetchStories])

  const handleRemoveStory = (item) => {
    // const newStories = stories.filter((story) => item.objectID !== story.objectID);
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item
    });
    // setStories(newStories);

    // dispatchStories({
      // type: 'SET_STORIES',
      // payload: newStories
    // });
  }

  const handleSearchInput = (event) => {
    setSearchTerm(event.target.value);
  };

  const handleSearchSubmit = (event) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);
    event.preventDefault();
  };

  /*const searchedStories = stories.data.filter((story) => story.title.toLowerCase().includes(searchTerm.toLowerCase())
  )*/

  return (
  <div className="App">
    <h1>{welcome.greetings} {welcome.title}</h1>

    <SearchForm
      searchTerm={searchTerm}
      onSearchInput={handleSearchInput}
      onSearchSubmit={handleSearchSubmit}
    />

    <hr />
    
    {stories.isError && <p>Something went wrong ...</p>}

    {stories.isLoading ?(

      <p>loading ...</p>
    ) : (

    // < List list={searchedStories} onRemoveItem={handleRemoveStory} />
    < List list={stories.data} onRemoveItem={handleRemoveStory} />
    )}
  </div>
)};

const SearchForm = ({
  searchTerm,
  onSearchInput,
  onSearchSubmit,
  }) => (
    <form onSubmit={onSearchSubmit}>
    <InputWithLabel 
      id="search" 
      isFocused 
      value={searchTerm} 
      // onInputChange={handleSearch}
      onInputChange={onSearchInput}
    >
      <strong>Search:</strong>
    </InputWithLabel>

    <button
      type="submit"
      disabled={!searchTerm}
      // onClick={handleSearchSubmit}
    >
    Submit
    </button>
  </form>
);

const InputWithLabel = ({id, type="text", isFocused, children, value, onInputChange}) => {

  const inputRef = React.useRef();

  React.useEffect(() => {
      if(isFocused && inputRef.current){

        inputRef.current.focus();
      }
    }, [isFocused]);

  return(
    <>
      <label htmlFor={id}>{children}:</label>
      <input ref={inputRef} type={type} value={value} onChange={onInputChange} />
    </>
  )
}

const List = ({list, onRemoveItem}) => (
  <ul>
    {list.map((item) => (
      <Item key={item.objectID} item = {item} onRemoveItem={onRemoveItem}/>
    ))}
  </ul>
  );

  const Item = ({item, onRemoveItem}) => {

    return(
    <li>
    <span>
      <a href={item.url}>{item.title}</a>  
    </span>
    <span>{item.author}</span>
    <span>{item.num_comments}</span>
    <span>{item.points}</span>
    <span>
      <button type="button" onClick={()=>onRemoveItem(item)}>Dismiss</button>
    </span>
    </li>
  )};

export default App;
