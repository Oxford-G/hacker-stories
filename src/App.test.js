import * as React from 'react';
import axios from 'axios';
jest.mock('axios');
import App, {
  storiesReducer,
  Item,
  List,
  SearchForm,
  InputWithLabel,
  } from './App';

  import {
    render,
    screen,
    fireEvent,
    act,
  } from '@testing-library/react';

const storyOne = {
  title: 'React',
  url: 'https://reactjs.org/',
  author: 'Jordan Walke',
  num_comments: 3,
  points: 4,
  objectID: 0,
};

const storyTwo = {
  title: 'Redux',
  url: 'https://redux.js.org/',
  author: 'Dan Abramov, Andrew Clark',
  num_comments: 2,
  points: 5,
  objectID: 1,
};

const stories = [storyOne, storyTwo];

describe('storiesReducer', () => {
  test('removes a story from all stories', () => {
    const action = { type: 'REMOVE_STORY', payload: storyOne };

    const state = { data: stories, isLoading: false, isError: false };

    const newState = storiesReducer(state, action);

    const expectedState = {
    data: [storyTwo],
    isLoading: false,
    isError: false,
    };

    expect(newState).toStrictEqual(expectedState);
  });

  test('initial loading of the app', ()=> {
    const action = {type: 'STORIES_FETCH_INIT'};

    const state = {data:[], isLoading: false, isError: false};

    const newState = storiesReducer(state, action);

    const expectedState = {data: [], isLoading: true, isError: false};

    expect(newState).toStrictEqual(expectedState)
  });

  test('succesful fetch', ()=> {
    const action = {type: 'STORIES_FETCH_SUCCESS', payload: stories};

    const state = {data: [], isLoading: false, isError: false};

    const newState = storiesReducer(state, action);

    const expectedState = {data: stories, isLoading: false, isError: false};

    expect(newState).toStrictEqual(expectedState)
  })

  test('failed fetch', ()=> {
    const action = {type: 'STORIES_FETCH_FAILURE'}

    const state = {data: [], isLoading: false, isError: false};

    const newState = storiesReducer(state, action)

    const expectedState = {data: [], isLoading: false, isError: true};

    expect(newState).toStrictEqual(expectedState)
  })
});

describe('Item', ()=> {
  test('renders all properties', ()=> {
    render(<Item item={storyOne} />)

    expect(screen.getByText('Jordan Walke')).toBeInTheDocument()
    expect(screen.getByText('React')).toHaveAttribute('href', 'https://reactjs.org/')
  });

  test('renders a clickable dismiss button', ()=> {
    render(< Item item={stories}/>)

    expect(screen.getByRole('button')).toBeInTheDocument()
  })

  test('clicking the dismiss button calls the callback handler', ()=> {
    const handleRemoveItem = jest.fn();

    render(< Item item={stories} onRemoveItem={handleRemoveItem}/>);
    fireEvent.click(screen.getByRole('button'));
    expect(handleRemoveItem).toHaveBeenCalledTimes(1)
  })
});

describe('SearchForm', ()=> {
  const searchFormProps = {
    searchTerm: 'React',
    onSearchInput: jest.fn(),
    onSearchSubmit: jest.fn(),
  };

  test('renders the input field with its value', ()=> {
    render(<SearchForm {...searchFormProps}/>)
    expect(screen.getByDisplayValue('React')).toBeInTheDocument()
  })

  // test('renders the correct label', () => {
  //   render(<SearchForm {...searchFormProps} />);
  //   expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
  // });

  test('calls onSearchInput on input field change', ()=> {
    render(<SearchForm {...searchFormProps} />);
    fireEvent.change(screen.getByDisplayValue('React'), {
      target: {value: 'Redux'}
    })
    expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1)
  })

  test('calls onSearchSubmit on button submit click', () => {
    render(<SearchForm {...searchFormProps} />);
    fireEvent.submit(screen.getByRole('button'));
    expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
  });

  test('renders snapshot', ()=> {
    const {container} = render(<SearchForm {...searchFormProps} />)
    expect(container.firstChild).toMatchSnapshot()
  })
})

describe('List', ()=> {
  test('render all property', ()=> {
    render(<List list={stories}/>)
    expect(screen.getByText('Jordan Walke')).toBeInTheDocument()
    expect(screen.getByText('Dan Abramov, Andrew Clark')).toBeInTheDocument()
    expect(screen.getByText('React')).toHaveAttribute('href', 'https://reactjs.org/')
    expect(screen.getByText('Redux')).toHaveAttribute('href', 'https://redux.js.org/')
    
  });

  test('get the number of buttons in list', ()=> {
    render(< List list={stories}/>)

    expect(screen.getAllByRole('button').length).toBe(2)
    
  })
})

describe('App', () => {
  test('succeeds fetching data', async () => {
    const promise = Promise.resolve({
      data: {
        hits: stories,
      },
    });

    axios.get.mockImplementationOnce(() => promise);
    render(<App />);

  // expect(screen.queryByText(/Loading/)).toBeInTheDocument();

    await act(() => promise);
   
    // expect(screen.queryByText(/Loading/)).toBeNull();

    expect(screen.getByText('React')).toBeInTheDocument();
    expect(screen.getByText('Redux')).toBeInTheDocument();
    expect(screen.getAllByText('delete.svg').length).toBe(2);
  });

  test('fails fetching data', async () => {
    const promise = Promise.reject();
    axios.get.mockImplementationOnce(() => promise);
    render(<App />);
    
    // expect(screen.getByText(/Loading/)).toBeInTheDocument();

    try {
      await act(() => promise);
    } catch (error) {
      // expect(screen.queryByText(/Loading/)).toBeNull();
      expect(screen.queryByText(/went wrong/)).toBeInTheDocument();
    }
  });

  test('removes a story', async ()=> {
    const promise = Promise.resolve({
      data: {hits: stories}
    })

    axios.get.mockImplementationOnce(() => promise);
    render(<App />);
    await act(() => promise);
    expect(screen.getAllByText('delete.svg').length).toBe(2);
    expect(screen.getByText('Jordan Walke')).toBeInTheDocument();

    fireEvent.click(screen.getAllByText('delete.svg')[0]);

    expect(screen.getAllByText('delete.svg').length).toBe(1);
    expect(screen.queryByText('Jordan Walke')).toBeNull();
  });

  test('searches for specific stories', async ()=> {
    const reactPromise = Promise.resolve({
      data: {hits: stories}
    });

    const anotherStory = {
      title: 'JavaScript',
      url: 'https://en.wikipedia.org/wiki/JavaScript',
      author: 'Brendan Eich',
      num_comments: 15,
      points: 10,
      objectID: 3,
    };

    const javascriptPromise = Promise.resolve({
      data: {
        hits: [anotherStory],
      },
    });

    axios.get.mockImplementation((url) => {

      if (url.includes('React')) {
      return reactPromise;
      }

      if (url.includes('JavaScript')) {
        return javascriptPromise;
      }

      throw Error();
    })

    // Initial Render

    render(<App />);

    // First Data Fetching

    await act(() => reactPromise);

    expect(screen.queryByDisplayValue('React')).toBeInTheDocument();
    expect(screen.queryByDisplayValue('JavaScript')).toBeNull();

    expect(screen.queryByText('Jordan Walke')).toBeInTheDocument();

    expect(screen.queryByText('Dan Abramov, Andrew Clark')).toBeInTheDocument();
    expect(screen.queryByText('Brendan Eich')).toBeNull();

      // User Interaction -> Search

    fireEvent.change(screen.queryByDisplayValue('React'), {
      target: {
      value: 'JavaScript',
      },
    });

    expect(screen.queryByDisplayValue('React')).toBeNull();
    expect(screen.queryByDisplayValue('JavaScript')).toBeInTheDocument();

    fireEvent.submit(screen.queryByText('Submit'));

    // Second Data Fetching

    await act(() => javascriptPromise);
    expect(screen.queryByText('Jordan Walke')).toBeNull();
    expect(screen.queryByText('Dan Abramov, Andrew Clark')).toBeNull();
    expect(screen.queryByText('Brendan Eich')).toBeInTheDocument();
  });
});
