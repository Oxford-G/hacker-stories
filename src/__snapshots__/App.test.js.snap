// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders snapshot 1`] = `
<form
  class="sc-iCfMLu eaaiGg"
>
  <label
    class="sc-furwcr jmfmdZ"
    for="search"
  >
    <strong>
      Search:
    </strong>
    :
  </label>
  <input
    class="sc-pVTFL gVFCYJ"
    type="text"
    value="React"
  />
  <button
    class="sc-eCImPb sc-gKclnd dQwsXJ dprkxb"
    type="submit"
  >
    Submit
  </button>
</form>
`;
